### PR TITLE
Use $PTHREAD_CFLAGS) on the relevant source file compilation and the library linking

### PR DESCRIPTION
**Pull request by ampli**
Regarding issue #1195, these are my proposed changes.

They don't set the global CFLAGS, but instead, use $(PTHREAD_CFLAGS) on the `tests` sources that use pthread, on all the library sources, and on the library linking.

I validated it by inspecting `make V=1` and ensuring that `-pthread` appears on these source file compilation and the library linking. It doesn't appear in the `link-parser` sources compilation and the python wrapper compilation (because it is not set in the global `CFLAGS` as I removed this line from `configure`).

Please test.

### CODE REVIEW

**Review by linas — APPROVED** [review]
Looks right to me. I'll test shortly and push version 5.9.1 in a few hours.

**Comment by linas** [comment]
This is insufficient to fix issue #1195  -- My testing procedure is this:
```
sudo apt remove libsqlite3-dev libaspell-dev
./autogen.sh --no-configure
cd build
rm -r link-parser link-grammar tests
../configure --disable-python-bindings --disable-java-bindings --disable-editline
make -j
ldd -r ./link-grammar/.libs/liblink-grammar.so.5.9.0
```
I still get
```
undefined symbol: tss_set	(./link-grammar/.libs/liblink-grammar.so.5.9.0)
undefined symbol: tss_create	(./link-grammar/.libs/liblink-grammar.so.5.9.0)
undefined symbol: call_once	(./link-grammar/.libs/liblink-grammar.so.5.9.0)
```
and `make check` fails.

**Comment by ampli** [comment]
I will do it exactly as you did and will investigate what is missing.

**Comment by linas** [comment]
This fixes it:
```
--- a/link-grammar/Makefile.am
+++ b/link-grammar/Makefile.am
@@ -74,6 +74,9 @@ endif WITH_SAT_SOLVER
 # Needed for tss_create and tss_set
 liblink_grammar_la_LIBADD  += ${PTHREAD_LIBS}
 
+# Brute-force fix for bug #1195
+liblink_grammar_la_LIBADD  += -lpthread
+
 # Math libraries are needed for floorf, etc.
 liblink_grammar_la_LIBADD  += -lm
```

I spent a few hours screwing around with `AX_PTHREAD` and other suggestive things in `/usr/share/aclocal` and had no luck. It appears that `${PTHREAD_LIBS}` is always empty, It also seems that linking with the `-pthread` flag does nothing discernible; it does not magically find the needed library.

I could just push the above brute-force fix, and issue 5.6.1 and cross my fingers that it doesn't wreck Apple or BSD or Windows....

**Comment by linas** [comment]
https://stackoverflow.com/questions/2127797/significance-of-pthread-flag-when-compiling

**Comment by linas** [comment]
Uh, hang on .. I failed to `git pull` your fixes ...  That's ... unexpected so ignore comments above, let me try again.

**Comment by linas** [comment]
... I now pulled your fixes .. and it still doesn't work. I'm going to bed.

**Comment by linas** [comment]
I can confirm the stackoverflow suggestion above does NOT end up using `-lpthread`. Look at `gcc -dumpspecs | less` and back up by one line to see what it applies to.

**Comment by ampli** [comment]
This is what I found:

 The `gcc` manual says:
`       -pthread
           Link with the POSIX threads library.  This option is supported on GNU/Linux targets, most other Unix derivatives, and also on x86 Cygwin and MinGW targets.  On some targets this option also sets flags for the preprocessor, so it should be
           used consistently for both compilation and linking.`
So it looks like that adding `-pthred` should cause it to link with `-lpthread`.

However, the link command line contains -pthread (even twice...) and it definitely doesn't link with `-lpthread` (I tried with a single `-lpthread` just to be sure).

The link command is:
`g++ -fPIC -DPIC -shared -nostdlib ... -pthread -Wl,-soname -Wl,liblink-grammar.so.5 -o .libs/liblink-grammar.so.5.9.0`
When I tried to remove the `-nostdlib` argument (plush removing the `crtX.o` stuff) the LG library got linked with the pthread library!

From googling `libtool pthread`,  it turned out this is an old and common problem with apparently no good solution, which may be considered as a bug in `gcc` (or by others as a bug in `libtool`).
See for example https://gcc.gnu.org/bugzilla/show_bug.cgi?id=25460:
`Because -nostdlib currently has priority over -pthread, libtool can't build shared C++ libraries using libpthread. A libtool maintainer told me to file a bug with GCC. Here's my original email: ...`
(He talks about c++, but I validated that this happens also when using `gcc` instead of `g++`.)

From reading various items in the said google search results, it looks like people are solving this problem by adding `-lpthread` when linking (`-pthread` is still needed when compiling to get `gcc` pthread support). But maybe there is a better solution that I haven't located.

In a hope of better portability, I propose this solution:
1. Use `PTHREAD_LIBS` on linking (it should have been used anyway). On Linux it is empty.
2. In `configure.ac`, if `PTHREAD_LIBS` is empty, assign to it `-lpthread`.

I can send a PR that implements it (of course after I test it.) But only later today.

BTW, what I still didn't find is why the linking of `link-parser` gets the `-pthread` flag (which causes overlinking).

**Comment by linas** [comment]
> I propose this solution:

Yes, that is the right solution.

FYI, the future is in flux: 
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=55394   -- removes dependency on pthread, just 6 months ago ...
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=99341 -- above is reverted  2 months ago.

This is why I think your proposed fix is the right one.

**Comment by linas** [comment]
One pill makes you larger, and the other pill makes you small: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66146
